### PR TITLE
Switch to fb webdriver

### PR DESCRIPTION
**Pull request by uuf6429**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ cache:
   directories:
     - $HOME/.composer/cache/files
 
-php: [5.3, 5.4, 5.5, 5.6, 7.0, 7.1, hhvm]
```
Comment (aik099) [inline .travis.yml]: Revert. See https://github.com/minkphp/MinkSelenium2Driver/issues/262#issuecomment-277508479.

File: .travis.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ before_script:
   # Start a webserver for web fixtures. Force using PHP 5.6 to be able to run it on PHP 5.3 and HHVM jobs too
   - MINK_PHP_BIN=~/.phpenv/versions/5.6/bin/php vendor/bin/mink-test-server > /dev/null 2>&1 &
 
-script: phpunit -v --coverage-clover=coverage.clover
```
Comment (aik099) [inline .travis.yml]: Revert. See https://github.com/minkphp/MinkSelenium2Driver/issues/262#issuecomment-277508548.

File: composer.json [hunk 3]
```diff
@@ -20,13 +20,14 @@
     ],
 
     "require": {
-        "php":                       ">=5.3.1",
```
Comment (aik099) [inline composer.json]: Revert. See https://github.com/minkphp/MinkSelenium2Driver/issues/262#issuecomment-277508479.

File: composer.json [hunk 4]
```diff
@@ -20,13 +20,14 @@
...
-        "behat/mink":                "~1.7@dev",
-        "instaclick/php-webdriver":  "~1.1"
+        "php":                   ">=5.5",
+        "behat/mink":            "~1.7@dev",
+        "facebook/webdriver":    "~1.3"
     },
 
     "require-dev": {
-        "mink/driver-testsuite": "dev-master"
```
Comment (aik099) [inline composer.json]: Revert. See https://github.com/minkphp/MinkSelenium2Driver/issues/262#issuecomment-277508548.

File: phpunit.xml.dist [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
-<phpunit colors="true" bootstrap="vendor/autoload.php">
```
Comment (aik099) [inline phpunit.xml.dist]: Revert. You can specify it for now in `.travis.yml` and command line option to PHPUnit. No need to alter default phpunit config file.

File: src/Selenium2Driver.php [hunk 6]
```diff
@@ -68,25 +70,31 @@ class Selenium2Driver extends CoreDriver
...
+     * @param string $browserName Browser name
+     * @param array $desiredCapabilities The desired capabilities
+     * @param string $wdHost The WebDriver host
      */
     public function __construct($browserName = 'firefox', $desiredCapabilities = null, $wdHost = 'http://localhost:4444/wd/hub')
     {
         $this->setBrowserName($browserName);
         $this->setDesiredCapabilities($desiredCapabilities);
-        $this->setWebDriver(new WebDriver($wdHost));
+        $this->wdHost = $wdHost;
```
Comment (aik099) [inline src/Selenium2Driver.php]: Any reason why we can't create WebDriver at this point as it was done before?

File: src/Selenium2Driver.php [hunk 7]
```diff
@@ -68,25 +70,31 @@ class Selenium2Driver extends CoreDriver
...
     /**
      * Sets the browser name
      *
      * @param string $browserName the name of the browser to start, default is 'firefox'
+     *
+     * @throws DriverException
      */
     protected function setBrowserName($browserName = 'firefox')
     {
+        if ($this->started) {
```
Comment (aik099) [inline src/Selenium2Driver.php]: How can somebody call protected method on a non-started session?

File: src/Selenium2Driver.php [hunk 8]
```diff
@@ -107,45 +115,12 @@ public function setDesiredCapabilities($desiredCapabilities = null)
         }
 
         if (null === $desiredCapabilities) {
-            $desiredCapabilities = array();
```
Comment (aik099) [inline src/Selenium2Driver.php]: Revert. Please don't use PHP syntax not compatible with PHP 5.3.

File: src/Selenium2Driver.php [hunk 9]
```diff
@@ -107,45 +115,12 @@ public function setDesiredCapabilities($desiredCapabilities = null)
...
+            $desiredCapabilities = [];
         }
 
         // Join $desiredCapabilities with defaultCapabilities
         $desiredCapabilities = array_replace(self::getDefaultCapabilities(), $desiredCapabilities);
 
-        if (isset($desiredCapabilities['firefox'])) {
```
Comment (aik099) [inline src/Selenium2Driver.php]: Any reasons for removing all that code? Maybe this is done differently now and if it is, then does related test pass?

File: src/Selenium2Driver.php [hunk 10]
```diff
@@ -160,19 +135,9 @@ public function getDesiredCapabilities()
     }
 
     /**
-     * Sets the WebDriver instance
-     *
-     * @param WebDriver $webDriver An instance of the WebDriver class
-     */
-    public function setWebDriver(WebDriver $webDriver)
```
Comment (aik099) [inline src/Selenium2Driver.php]: Maybe we can still allow that behavior somehow?

File: src/Selenium2Driver.php [hunk 11]
```diff
@@ -186,17 +151,17 @@ public function getWebDriverSession()
      */
     public static function getDefaultCapabilities()
     {
-        return array(
-            'browserName'       => 'firefox',
```
Comment (aik099) [inline src/Selenium2Driver.php]: No need to remove formatting of multi-line array calls.

File: src/Selenium2Driver.php [hunk 12]
```diff
@@ -240,27 +205,27 @@ protected static function charToOptions($char, $modifier = null)
...
         if ($modifier) {
-            $options[$modifier.'Key'] = 1;
+            $options[$modifier . 'Key'] = 1;
         }
 
         return json_encode($options);
     }
 
     /**
-     * Executes JS on a given element - pass in a js script string and {{ELEMENT}} will
```
Comment (aik099) [inline src/Selenium2Driver.php]: Why the `{{ELEMENT}}` was replaced with `__ELEMENT__`? If I recall correctly it's driver itself who parses `{{ELEMENT}}` and not WebDriver/Facebook.

File: src/Selenium2Driver.php [hunk 13]
```diff
@@ -298,15 +263,37 @@ private function executeJsOnElement(Element $element, $script, $sync = true)
...
      * {@inheritdoc}
      */
     public function start()
     {
         try {
-            $this->wdSession = $this->webDriver->session($this->browserName, $this->desiredCapabilities);
+            $this->wdSession = new Selenium2DriverSession($this->buildWebDriver());
             $this->applyTimeouts();
         } catch (\Exception $e) {
-            throw new DriverException('Could not open connection: '.$e->getMessage(), 0, $e);
```
Comment (aik099) [inline src/Selenium2Driver.php]: Revert. No need to change formatting.

File: src/Selenium2Driver.php [hunk 14]
```diff
@@ -762,9 +752,13 @@ public function click($xpath)
         $this->clickOnElement($this->findElement($xpath));
     }
 
+    /**
+     * @param Element $element
+     */
     private function clickOnElement(Element $element)
     {
-        $this->wdSession->moveto(array('element' => $element->getID()));
+        //$element->click(); // TODO why was this needed?
```
Comment (aik099) [inline src/Selenium2Driver.php]: What this comment is for? The `click` is called because method is called `clickOnElement` 😄

File: src/Selenium2Driver.php [hunk 15]
```diff
@@ -793,36 +787,23 @@ public function attachFile($xpath, $path)
...
-        // as a remote instance.
-        try {
-          $remotePath = $this->uploadFile($path);
-        } catch (\Exception $e) {
-          // File could not be uploaded to remote instance. Use the local path.
-          $remotePath = $path;
-        }
-
-        $element->postValue(array('value' => array($remotePath)));
+        $element->sendKeys($path);
```
Comment (aik099) [inline src/Selenium2Driver.php]: Are you sure, that `sendKeys` is clever enough to handle file upload code on local/remote servers?

File: phpunit.xml.dist [hunk 16]
```diff
@@ -1,6 +1,6 @@
...
+<phpunit colors="true" bootstrap="vendor/autoload.php" stopOnFailure="true">
```
Comment (uuf6429) [inline phpunit.xml.dist]: This is temporary until we get all tests working.

File: src/Selenium2Driver.php (hunk 9 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Seems that the facebook webdriver already does this process... I have to be sure though.

File: src/Selenium2Driver.php (hunk 12 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: 1. It shows up as JS errors in my IDE
2. It seems to be replaced by webdriver: https://github.com/uuf6429/MinkSelenium2Driver/blob/05872053bd733959b25fa4df3f3aa41bef845d64/src/Selenium2Driver.php#L251-L251

File: src/Selenium2Driver.php (hunk 10 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Is this behaviour really required? To me it seems like the perfect invitation for hackeries and seriously weird bugs.

File: src/Selenium2Driver.php (hunk 7 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Ah, good point... didn't realise it's protected.

File: phpunit.xml.dist (hunk 16 above)
Comment (aik099) [inline phpunit.xml.dist]: Still please move it to `.travis.yml`. This would allow people checking out PR branch (for reviewing purposes) like me and using alternative PHPUnit runner not have this stopOnFailure behavior.

File: src/Selenium2Driver.php (hunk 6 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: If you look at the code of `buildWebdriver()`, you'll see that when the web driver is build, it requires the capabilities.
If I move that code up here, `setDesiredCapabilities()` will not work anymore.

File: src/Selenium2Driver.php (hunk 9 above)
Comment (aik099) [inline src/Selenium2Driver.php]: I'm not sure that we have tests to cover passing of browser-specific desired capabilities though. If we do, then they'll catch fact that something isn't passed.

File: src/Selenium2Driver.php (hunk 15 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Yes: https://github.com/facebook/php-webdriver/blob/85aa68020ef2cf4fd261b1f1a690052afe67a9e0/lib/Remote/RemoteWebElement.php#L324

File: src/Selenium2Driver.php (hunk 10 above)
Comment (aik099) [inline src/Selenium2Driver.php]: Looks like that. I'll we'll address that later on.

File: src/Selenium2Driver.php (hunk 6 above)
Comment (aik099) [inline src/Selenium2Driver.php]: That will do for now then. Thanks for explaining.

File: src/Selenium2Driver.php (hunk 14 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: It's called `clickOnElement()`, not `doubleClickOnElement()`... `$element->click()` was called twice. What's the reason?

Comment (uuf6429) [inline src/Selenium2Driver.php]: Ah, never mind... seems to be an artefact from my changes.

File: src/Selenium2Driver.php (hunk 7 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Actually, I just realized... the point of a protected method is that it can be called in a child class, no?
So if someone calls it from child class, then the validation is still valid. If not, then we don't need this method in the first place and we can as well assign the property in the constructor.

File: src/Selenium2Driver.php (hunk 12 above)
Comment (aik099) [inline src/Selenium2Driver.php]: Of course it will show as JavaScript error in IDE because `{{ELEMENT}}` is invalid syntax. Since it works please replace `__ELEMENT__` back into `{{ELEMENT}}`.

File: src/Selenium2Driver.php (hunk 7 above)
Comment (aik099) [inline src/Selenium2Driver.php]: Fact, that no exception was thrown before proves that nobody really does that.

If you'd like, then you can move exception throwing into separate PR since it's not related to switching to Facebook's WebDriver in any way.

File: src/Selenium2Driver.php (hunk 15 above)
Comment (aik099) [inline src/Selenium2Driver.php]: Good. In any case we have functional tests to prove that it really works as expected.

File: src/Selenium2Driver.php (hunk 14 above)
Comment (aik099) [inline src/Selenium2Driver.php]: Then you can safely remove it.

File: src/Selenium2Driver.php (hunk 12 above)
Comment (uuf6429) [inline src/Selenium2Driver.php]: Why do you want to show errors when they're in fact just replaced tokens? O.o

Comment (aik099) [inline src/Selenium2Driver.php]: Because changes in PR should be on topic. Making unrelated change doesn't help anybody.

Comment (aik099) [inline src/Selenium2Driver.php]: You can send another PR with just that change and we can merge/not merge it separately regardless of faith of this PR.

Comment (stof) [inline src/Selenium2Driver.php]: Plus, it is a BC break as it is a change affecting a protected method

**Comment by stof** [comment]
Why closing the PR ?

**Comment by aik099** [comment]
> Why closing the PR ?

Because nobody noticed upfront that Facebook's version of WebDriver requires PHP 5.5+ to run. Since Mink/Mink Driver require PHP 5.3+ to run we have no way to use it.

**Comment by stof** [comment]
Well, as PHP 5.3 and 5.4 are EOLed, dropping support for them could be acceptable if we have some benefit when doing it (switching to a maintained library).
The previous PR dropping it for Mink was dropping support with no benefit for us. Here, it could have one.

Some drivers may drop support for older PHP versions (the third-party phantomJS driver does not support PHP 5.3 btw)

File: src/Selenium2DriverSession.php [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+
+namespace Behat\Mink\Driver;
+
+use Behat\Mink\Exception\DriverException;
+use Facebook\WebDriver\Remote\DriverCommand;
+use Facebook\WebDriver\Remote\RemoteWebDriver;
+use Facebook\WebDriver\WebDriverBy;
+use Facebook\WebDriver\WebDriverElement;
+
+class Selenium2DriverSession
```
Comment (stof) [inline src/Selenium2DriverSession.php]: Why creating this class ?

**Comment by aik099** [comment]
> Here, it could have one.

Switching one dependency for another, while both work at the end is no benefit to me. Fact that one dependency is better maintained than another one doesn't really change the fact that both work.

File: src/Selenium2DriverSession.php (hunk 17 above)
Comment (uuf6429) [inline src/Selenium2DriverSession.php]: It's a proxy class to avoid changing the original driver too much. In the end, I could merge all changes into one class.

**Comment by bwowk** [comment]
@aik099 actually there are many limitations on the instaclick version that are not present on the facebook one. Off the top of my head I can remember the iframe manipulation bit, for instance. On instaclick/php-webdriver you can only change iframes based on it's id, whereas in facebook/php-webdriver you can use the id, name or an WebDriverElement

**Comment by aik099** [comment]
> Off the top of my head I can remember the iframe manipulation bit, for instance.

This particular change isn't a problem, because public API of Mink supporting locating frames by name only anyway.

Since Instaclick version covers 100% of functionality needed to implement the driver it doesn't really matter at this point if we switch to FB at all.

**Comment by uuf6429** [comment]
Some other reasons:
- maintained library is possibly faster and in case of performance/bug/security issues, there's a higher chance of a patch
- the FB library takes care of things like handling uploads and other stuff that is unnecessarily replicated in MinkSelenium2Driver.

That said, I gave up arguing here. It's a huge shame this project will die like this "because it works".

**Comment by bwowk** [comment]
@uuf6429 no reason for it to die at all, just put your fork up on Packagist and I'll help you maintain it. No trouble at all if the main one is stagnant and becomes obsolete, we'll just move forward using yours.

**Comment by aik099** [comment]
I see you're eager to change stuff even though it works as expected for years now. I'm not here to judge how you should spend your time, but in your place I won't be spending time on doing anything that won't result in any kind of gain.

In this case I don't see any actual gains except `FB library is more maintained than Instaclick` one. There can't be any possible performance gain, because both FB and Instaclick are doing remote calls using WebDriver protocol, where speed is result of how fast we get a response and not how fast library can process it.

**Comment by Einenlum** [comment]
Seems instaclick does not officially support Selenium 3.x, am I right?
Therefore, if a library is still active (facebook) and the other is not, plus does not supports officially Selenium 3.x, why shouldn't we change this dependency?

Moreover, if it's a question of BC and minimum PHP version, it's also our duty to make all this community evolve. Even 5.6 is not actively supported anymore. People with previous PHP versions will still have older versions of MinkSelenium2Driver.
